src/page/mod.rs

```diff
@@ -45,41 +46,62 @@ pub trait Materializer: Send + Sync + Clone {
     /// Used to feed custom recovery information back to a higher-level abstraction
     /// during startup. For example, a B-Link tree must know what the current
     /// root node is before it can start serving requests.
-    fn recover(&mut self, &Self::PartialPage) -> Option<Self::Recovery>;
+    fn recover(&self, &Self::PartialPage) -> Option<Self::Recovery>;
 }
 
 /// Points to either a memory location or a disk location to page-in data from.
 #[derive(Clone, PartialEq)]
-pub enum CacheEntry<M>
-    where M: Materializer
-{
+pub enum CacheEntry<M> {
     /// A cache item that is in memory, and also in secondary storage.
-    Resident(Vec<M::PartialPage>, LogID),
+    Resident(Vec<M>, LogID),
     /// A cache item that is present in secondary storage.
     PartialFlush(LogID),
     /// A cache item that is present in secondary storage, and is the base segment
     /// of a page.
     Flush(LogID),
 }
 
-unsafe impl<M: Materializer> Send for CacheEntry<M> {}
+unsafe impl<M: Send> Send for CacheEntry<M> {}
 
 /// `LoggedUpdate` is for writing blocks of `Update`'s to disk
 /// sequentially, to reduce IO during page reads.
+#[serde(bound(deserialize = ""))]
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-struct LoggedUpdate<M>
-    where M: Materializer
+pub(super) struct LoggedUpdate<PartialPage>
+    where PartialPage: Serialize + DeserializeOwned
 {
     pid: PageID,
-    update: Update<M>,
+    update: Update<PartialPage>,
 }
 
+#[serde(bound(deserialize = ""))]
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-enum Update<M>
-    where M: Materializer
+enum Update<PartialPage>
+    where PartialPage: DeserializeOwned + Serialize
 {
-    Append(Vec<M::PartialPage>),
-    Compact(Vec<M::PartialPage>),
+    Append(Vec<PartialPage>),
+    Compact(Vec<PartialPage>),
     Del,
     Alloc,
 }
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+struct Snapshot<R> {
+    pub max_lid: LogID,
+    pub max_pid: PageID,
+    pub pt: BTreeMap<PageID, Vec<LogID>>,
+    pub free: Vec<PageID>,
+    pub recovery: Option<R>,
+}
```

Comment:
This is the main structure that gets written to disk when we take a snapshot. Note that we store only a Vec of LogID's, which is a pointer to where to find the data later, rather than any actual data.